Append code coverage checks in CI pipeline

Rolland0000 opened the pull request:
Issues #26
- Add code coverage job in the CI pipeline
- Use docker image of Tarpaulin
- Set the covrage threshold at 70%

RollandAyonta committed:
ci(workflow): Enhance CI pipeline with code coverage and branch targeting

- Add code coverage job using tarpaulin
- Implement 70% coverage threshold check
- Upload coverage report as artifact
- Configure workflow to trigger on push and main branch pull requests
- Add Windows MSVC target configuration for mold/lld linker in .cargo/config.toml

Rolland0000 committed:
chore(config): Disable custom linker and rustflags configurations

- Comment out platform-specific linker and rustflags settings

Rolland0000 committed:
ci(coverage): Improve code coverage threshold validation

- Fix code coverage threshold check in CI workflow

Rolland0000 committed:
ci(coverage): Improve code coverage reporting and threshold validation

- Modify coverage job to capture and display full tarpaulin output

Rolland0000 committed:
ci(dependencies): Disable cargo-machete metadata check

Hermann-Core commented on .cargo/config.toml:
```diff
@@ -1,6 +1,9 @@
-[target.x86_64-apple-darwin]
-rustflags = ["-C", "link-arg=-fuse-ld=lld"]
+# [target.x86_64-apple-darwin]
+# rustflags = ["-C", "link-arg=-fuse-ld=lld"]
```
Why are you commenting these out?

Hermann-Core commented on .github/workflows/ci.yml:
```diff
@@ -1,8 +1,10 @@
 name: CI Pipeline
 
 on:
+  push:
```
Dou you need to run the pipeline on every push?

Hermann-Core commented on .github/workflows/ci.yml:
```diff
@@ -40,8 +42,41 @@ jobs:
         with:
           tool: cargo-machete
-          args: check --with-metadata
+          # args: check --with-metadata
```
Why did you comment this out?

Hermann-Core commented on .github/workflows/ci.yml:
```diff
@@ -40,8 +42,41 @@ jobs:
+      
+      - name: Generate coverage with threshold check
+        run: |
+          cargo tarpaulin --verbose --workspace --out xml --out html --output-dir ./coverage > coverage_output.txt 2>&1
+          cat coverage_output.txt
+          
+          COVERAGE=$(grep -oP '\d+\.\d+(?=% coverage)' coverage_output.txt | head -1)
+          
+          if [ -z "$COVERAGE" ]; then
+            echo "Failed to extract coverage"
+            exit 1
+          fi
+          
+          echo "Coverage: ${COVERAGE}%"
+          if (( $(echo "$COVERAGE < 70" | bc -l) )); then
+            exit 1
+          fi
+      
+      - name: Upload coverage artifact
+        uses: actions/upload-artifact@v4
+        if: always()
+        with:
+          name: coverage-report
+          path: coverage/
```
Could you check and reuse the same standard we are using for the other steps?

Hermann-Core commented on .gitignore:
```diff
@@ -1 +1,2 @@
 target/
+.cargo\config.toml
```
This file should not be removed from the version control

Rolland0000 replied:
okay

Rolland0000 committed:
ci(config): Enable custom linker and simplify CI pipeline

- Re-enable clang linker and mold configuration for Linux and Darwin targets
- Remove Windows MSVC linker configuration
- Simplify CI workflow to run on pull requests only
- Replace container-based tarpaulin with standard setup-rust-toolchain approach
- Add explicit cargo machete execution step with metadata check
- Streamline coverage job by removing dependency on ci-checks job
- Update cargo-tarpaulin installation to use install-action
- Simplify coverage threshold validation using built-in fail-under flag
- Fix .gitignore to properly exclude .cargo/config.toml
- Improve CI pipeline maintainability and reduce complexity

Rolland0000 committed:
ci(workflow): Simplify CI pipeline step formatting

- Remove unnecessary step name from cargo machete task
- Consolidate cargo tarpaulin command into single line for readability

Rolland0000 committed:
ci(coverage): Replace tarpaulin with llvm-cov for code coverage

- Replace cargo-tarpaulin with cargo-llvm-cov for improved coverage reporting
- Add nextest installation to CI workflow for better test execution
- Maintain 70% code coverage threshold with llvm-cov's --fail-under-lines flag
- Add lcov.info file for coverage data tracking
- Add health check integration tests in tests/test_health_check.rs
- Update Cargo.lock and Cargo.toml with new dependencies and version bumps

Rolland0000 committed:
ci(coverage): Remove lcov output format from coverage generation

- Remove --lcov flag from cargo llvm-cov
